### DISCUSSION TITLE
Manually specifying root file in project with multiple root files

### DISCUSSION THREAD

**Discussion by mhthies** [James-Yu/LaTeX-Workshop#3172] [Q&A]
Hi,

I'd like to know if there is a mechanism for manually specifying the root file to be built and viewed. I often work on LaTeX projects with mutliple root files that share included files within the same project.

- One of my use cases is building different versions of the same document: The different root files make different defines, before including the same "main" document file (which acually comprises the `\begin{document}` then).
- Another use case is creating university exams in LaTeX, where I have one document for each semester and a pool of exercises to be included by the exam documents. Each exercise may be included by multiple exams.

I even combine these two use cases, e.g. to create an exam document and a sample solution for each semester's exam.

I know about #12, but that's a rather old issue, so I wanted to ask here, if a solution has been added since then.

**Comment by jlelong** [comment]
> * One of my use cases is building different versions of the same document: The different root files make different defines, before including the same "main" document file (which actually comprises the `\begin{document}` then).

Here, your issue is that the root file does not contain the `\begin{document}`. I would advise to do it the other way round. At least, this is how I deal with the same use case as yours.
- Remove the `\begin{document}` from your current _main_ document
- In the files to be root files, put something like
    ```latex
    \documentclass{article}
    some preamble stuff making the right definitions
    \begin{document}
    \input{one_of_your_current_main_files.tex}
    \end{document}
    ```
    


> * Another use case is creating university exams in LaTeX, where I have one document for each semester and a pool of exercises to be included by the exam documents. Each exercise may be included by multiple exams.

There are basically two solutions to this. 

- Use magic comments to define in every exercice which root file you want to consider, see https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#the-root-file. This can be tedious, hence the second solution, which is more of a trick actually
- Open the file you want to consider as a the root file, or visit the file it is already opened. The root file will not change as long as you keep editing files which are included in it. Note that this approach is somehow a side effect of the current implementation and may change without notice. Until then, you can use this mechanism to somehow set which root file you want to consider.

**Reply by mhthies** [reply]
> - In the files to be root files, put something like
>   ```tex
>   \documentclass{article}
>   some preamble stuff making the right definitions
>   \begin{document}
>   \input{one_of_your_current_main_files.tex}
>   \end{document}
>   ```

Hm, that doesn't really fit my use case. This "some preamble stuff" is a large part of the current main document, which shall be shared between the different document versions (= root files). I know, I could extract all of that into another file and include from all root files. But this makes it quite complicated to figure out the document structure, especially for my collegues, who are not all that far into TeX.

> Use magic comments to define in every exercice which root file you want to consider

This would be quite tedious and error-prone, since one is usually only working on the newest exam document, such that all re-used exercise files need to have their magic comments adapted such that they will point the the newest exam file as root document. However, the trick with opening/visiting the root document should do it.

IMHO, all of this would be solved by adding a simple configuration option for setting the root document file, circumventing the whole root detection logic. I could set this option as a workspace setting in VS Code to manually define which file I want to have compiled, when I change TeX files in this worksapce.

If that is not how LaTeX-Workshop is meant to work, I understand that and I will probably switch back to Sublime Text + LaTeXTools or some other editor for these TeX documents.

**Reply by jlelong** [reply]
> IMHO, all of this would be solved by adding a simple configuration option for setting the root document file, circumventing the whole root detection logic. I could set this option as a workspace setting in VS Code to manually define which file I want to have compiled, when I change TeX files in this workspace.

I will think of it. In the meantime, you may find a solution by using the [external command build mechanism](https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#external-build-command), which completely bypasses the root file detection.
